4.3.1 Data Preprocessing Part 2

4.3 Exit Assessment

In [1]:
#installing pandas package
!pip install pandas

Part 1: Using data found in sales.csv, perform the following task
1 Standardize the Price column
2 Identify and remove outliers in the dataset

In [3]:
#importing pandas library
import pandas as pd

#read sales.csv
data = pd.read_csv("C:/Users/mneme/Desktop/sales.csv")

#Checking null values
data.isnull().sum()

Transaction_date     3
Product              3
Price                1
Payment_Type         3
Name                 0
City                 4
State               13
Country              0
Account_Created      3
Last_Login           7
Latitude             3
Longitude            1
dtype: int64

Since, price has only 1 null values, deleting it wont affect price standardization

In [18]:
#Deleting null values in Price column
data.dropna(subset=['Price'], inplace = True)

Starndardization Formula
Xnew = (Xold - Xmean)/Standard Deviation

In [20]:
data.insert(12, "Standardized_Price", 0)

In [23]:
#importing numpy library
import numpy as np

#Removing comma in Price column
data['Price'] = data['Price'].replace(',','', regex = True)

#Converting string to float
data['Price'] = data['Price'].astype(float)

#Standardization of the Price
data.Standardized_Price = (data.Price - np.mean(data.Price))/(np.std(data.Price))

In [24]:
print(data)

    Transaction_date   Product   Price Payment_Type               Name  \
0      1/2/2009 6:17  Product1  1200.0   Mastercard           carolina   
1      1/2/2009 4:53  Product1  1200.0         Visa             Betina   
2     1/2/2009 13:08  Product1  1200.0   Mastercard  Federica e Andrea   
3     1/3/2009 14:44  Product1  1200.0         Visa              Gouya   
4     1/4/2009 12:56  Product2  3600.0         Visa             Gerd W   
..               ...       ...     ...          ...                ...   
995  1/22/2009 14:25  Product1  1200.0         Visa         Hans-Joerg   
996   1/28/2009 5:36  Product2  3600.0         Visa         Christiane   
997    1/1/2009 4:24  Product3  7500.0         Amex             Pamela   
998   1/8/2009 11:55  Product1  1200.0       Diners              julie   
999  1/12/2009 21:30  Product1  1200.0         Visa              Julia   

               City             State         Country   Account_Created  \
0          Basildon           Englan

In [25]:
#Getting the mean and standard deviation
mean = np.mean(data.Price)
std = np.mean(data.Price)
lowerlimit = max(0,mean - 3*std)
upperlimit = mean + 3*std

#Outlier Identification and Removal
data_cleaned1 = data[data.Price >= lowerlimit]
data_cleaned2 = data[data.Price <= upperlimit]

In [26]:
data_cleaned2.describe(include="all")

Transaction_date   Product        Price Payment_Type   Name    City  \
count               978       978   981.000000          979    981     977   
unique              960         2          NaN            4    727     736   
top       1/5/2009 5:33  Product1          NaN         Visa  Sarah  London   
freq                  2       842          NaN          511     11      19   
mean                NaN       NaN  1534.169215          NaN    NaN     NaN   
std                 NaN       NaN   834.950840          NaN    NaN     NaN   
min                 NaN       NaN    20.000000          NaN    NaN     NaN   
25%                 NaN       NaN  1200.000000          NaN    NaN     NaN   
50%                 NaN       NaN  1200.000000          NaN    NaN     NaN   
75%                 NaN       NaN  1200.000000          NaN    NaN     NaN   
max                 NaN       NaN  3600.000000          NaN    NaN     NaN   

          State        Country Account_Created      Last_Login    Latitude  \
count       968            981             978             974  979.000000   
unique      202             56             957             954         NaN   
top     England  United States  1/18/2009 0:00  1/4/2009 20:01         NaN   
freq         84            455               3               2         NaN   
mean        NaN            NaN             NaN             NaN   39.116890   
std         NaN            NaN             NaN             NaN   19.433659   
min         NaN            NaN             NaN             NaN  -41.465000   
25%         NaN            NaN             NaN             NaN   36.013475   
50%         NaN            NaN             NaN             NaN   42.395830   
75%         NaN            NaN             NaN             NaN   51.083333   
max         NaN            NaN             NaN             NaN   64.837780   

         Longitude  Standardized_Price  
count   980.000000          981.000000  
unique         NaN                 NaN  
top            NaN                 NaN  
freq           NaN                 NaN  
mean    -41.097999           -0.096533  
std      67.592075            0.690059  
min    -159.485280           -1.347942  
25%     -87.861807           -0.372712  
50%     -73.722083           -0.372712  
75%       4.916667           -0.372712  
max     174.766667            1.610806

Inter-Quartile Range (IQR) = Q75 - Q25

In [27]:
IQR = np.percentile(data.Price, 75) - np.percentile(data.Price, 25)
print(IQR)

upperlimit2 = np.percentile(data.Price, 75) + 1.5*IQR
lowerlimit2 = max(0, np.percentile(data.Price, 25) - 1.5*IQR)
print(upperlimit2)
print(lowerlimit2)

0.0
1200.0
1200.0


In [28]:
data_cleaned3 = data[data.Price >= lowerlimit]
data_cleaned4 = data[data.Price <= upperlimit]

In [29]:
data_cleaned4.describe(include="all")

Transaction_date   Product        Price Payment_Type   Name    City  \
count               978       978   981.000000          979    981     977   
unique              960         2          NaN            4    727     736   
top       1/5/2009 5:33  Product1          NaN         Visa  Sarah  London   
freq                  2       842          NaN          511     11      19   
mean                NaN       NaN  1534.169215          NaN    NaN     NaN   
std                 NaN       NaN   834.950840          NaN    NaN     NaN   
min                 NaN       NaN    20.000000          NaN    NaN     NaN   
25%                 NaN       NaN  1200.000000          NaN    NaN     NaN   
50%                 NaN       NaN  1200.000000          NaN    NaN     NaN   
75%                 NaN       NaN  1200.000000          NaN    NaN     NaN   
max                 NaN       NaN  3600.000000          NaN    NaN     NaN   

          State        Country Account_Created      Last_Login    Latitude  \
count       968            981             978             974  979.000000   
unique      202             56             957             954         NaN   
top     England  United States  1/18/2009 0:00  1/4/2009 20:01         NaN   
freq         84            455               3               2         NaN   
mean        NaN            NaN             NaN             NaN   39.116890   
std         NaN            NaN             NaN             NaN   19.433659   
min         NaN            NaN             NaN             NaN  -41.465000   
25%         NaN            NaN             NaN             NaN   36.013475   
50%         NaN            NaN             NaN             NaN   42.395830   
75%         NaN            NaN             NaN             NaN   51.083333   
max         NaN            NaN             NaN             NaN   64.837780   

         Longitude  Standardized_Price  
count   980.000000          981.000000  
unique         NaN                 NaN  
top            NaN                 NaN  
freq           NaN                 NaN  
mean    -41.097999           -0.096533  
std      67.592075            0.690059  
min    -159.485280           -1.347942  
25%     -87.861807           -0.372712  
50%     -73.722083           -0.372712  
75%       4.916667           -0.372712  
max     174.766667            1.610806

In [30]:
#export data to Edepcion_sales.csv file
data_cleaned4.to_csv("C:/Users/mneme/Desktop/Edepcion_sales.csv", index = False)

Part 2: Using the data in Handling Missing Values.csv
1 Handle all missing values. Provide an explanation for choosing a specific method (delete, impute or keep; if impute, why mean, median or mode?) in handling specific missing values

In [31]:
#read sales.csv
data_missing = pd.read_csv("C:/Users/mneme/Desktop/Handling_Missing_Values.csv")

#Checking null values
data_missing.isnull().sum()

ID          0
Sex         4
Age         0
Income      3
Employed    0
Children    2
Buy_Car     0
dtype: int64

Columns that has missing values are Sex = 4, Income = 3 and Children = 2

I will use imputation on all missing values

For Sex, since it is a categorical variable then I will replace the missing values with "No data" to preserve the original data

For Income, since it is numerical variable and in normal distribution, then I will use mean to replace the missing values

For Children, since it is numerical and it is skewed then I will use median to replace the missing values

In [33]:
#Getting the mode, mean and median
mean_value = data_missing['Income'].mean()
median_value = data_missing['Children'].median()

#Filling in the missing values with mode, mean and median
data_missing['Sex'] = data_missing['Sex'].fillna("No data")
data_missing['Income'] = data_missing['Income'].fillna(mean_value)
data_missing['Children'] = data_missing['Children'].fillna(median_value)
print(data_missing)

    ID      Sex  Age        Income Employed  Children Buy_Car
0    1     Male   25  25146.000000   Single       0.0      No
1    2     Male   30  26939.000000  Married       2.0     Yes
2    3     Male   27  26693.000000  Married       0.0      No
3    4     Male   28  26666.000000  Married       3.0     Yes
4    5     Male   29  25899.000000  Married       0.0      No
5    6     Male   28  26462.000000  Married       1.0      No
6    7   Female   28  26245.705882  Married       3.0     Yes
7    8  No data   30  26037.000000  Married       2.0     Yes
8    9   Female   28  26167.000000  Married       1.0     Yes
9   10  No data   28  26245.705882   Single       1.0      No
10  11   Female   28  26905.000000   Single       0.0      No
11  12     Male   30  26195.000000   Single       1.0     Yes
12  13   Female   30  26514.000000  Married       4.0     Yes
13  14   Female   28  26969.000000  Married       4.0     Yes
14  15  No data   28  26234.000000  Married       0.0      No
15  16  

In [34]:
#export data to Edepcion_Handling_Missing_Values.csv file
data_missing.to_csv("C:/Users/mneme/Desktop/Edepcion_Handling_Missing_Values.csv", index = False)

Part 3: 
1 Gather Data From Any Source (it would be better if you use your own data)
2 Preprocess the data you gathered

In [4]:
#read googleplaystore.csv, we will be using googleplaystore data for part 3
data_app = pd.read_csv("C:/Users/mneme/Desktop/googleplaystore.csv")

#initial view of the entire dataframe
data_app.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [6]:
data_app.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

We see that Rating has a lot of null values. So, we use imputation or fillna methods

In [7]:
data_app['Rating'] = data_app['Rating'].fillna(method = 'bfill' , axis=0).fillna(method = 'ffill' , axis=0)
data_app.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              1
Price             0
Content Rating    1
Genres            0
Last Updated      0
Current Ver       8
Android Ver       3
dtype: int64

In [8]:
data_app[data_app.Size == 'Varies with device'].shape

(1695, 13)

We find that the Size column has values of 1695 = 'Varies with device'. We will replace it by imputation or fillna. Also, removing special characters. And we make all values in the Size column to be numeric. 

In [12]:
#import numpy library
import numpy as np

data_app['Size'] = data_app['Size'].str.replace('+','')
data_app['Size'] = data_app['Size'].str.replace(',','')
data_app['Size'] = data_app['Size'].replace('Varies with device' , np.nan)
data_app['Size'] = data_app['Size'].str.replace('M' , '000000')
data_app['Size'] = data_app['Size'].str.replace('k' , '000')
data_app['Size'] = data_app['Size'].apply(lambda x: float(x)/1000)
data_app['Size'].dtype
data_app['Size'] = data_app['Size'].fillna(method = 'bfill' , axis=0).fillna(method = 'ffill' , axis=0)
data_app.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              1
Price             0
Content Rating    1
Genres            0
Last Updated      0
Current Ver       8
Android Ver       3
dtype: int64

In [14]:
data_app['Installs'].unique()

array(['10000', '500000', '5000000', '50000000', '100000', '50000',
       '1000000', '10000000', '5000', '100000000', '1000000000', '1000',
       '500000000', '50', '100', '500', '10', '1', '5', '0', 'Free'],
      dtype=object)

Removing unwanted characters in Installs. Deleting non-numeric values. Converting data type to int

In [22]:

data_app['Installs'] = data_app['Installs'].str.replace('+','')
data_app['Installs'] = data_app['Installs'].str.replace(',','')
data_app.drop(data_app[(data_app['Installs'] == 'Free')].index, inplace=True)
data_app[~data_app.Installs.str.isnumeric()]
data_app['Installs'] = data_app['Installs'].astype(int)
data_app['Installs'].dtype

dtype('int32')

In [24]:
#Checking Price column
data_app['Price'].unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$24.99', '$11.99', '$79.99', '$16.99', '$14.99',
       '$1.00', '$29.99', '$12.99', '$2.49', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', '$1.20', '$1.04'], dtype=object)

Removing unwanted characters in Price column. Converting data type to float.

In [25]:
data_app['Price'] = data_app['Price'].str.replace('$','')
data_app['Price'] = data_app['Price'].astype(float)
data_app['Price'].dtype

dtype('float64')

Handling the remaining missing values with imputation or fillna

In [26]:
data_app.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              1
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       8
Android Ver       2
dtype: int64

In [27]:
data_app['Type'] = data_app['Type'].fillna(method = 'bfill' , axis=0).fillna(method = 'ffill' , axis=0)
data_app['Current Ver'] = data_app['Current Ver'].fillna(method = 'bfill' , axis=0).fillna(method = 'ffill' , axis=0)
data_app['Android Ver'] = data_app['Android Ver'].fillna(method = 'bfill' , axis=0).fillna(method = 'ffill' , axis=0)
data_app.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

Converting Reviews column to numeric or int data type.

In [32]:
data_app['Reviews'].unique()

array(['159', '967', '87510', ..., '603', '1195', '398307'], dtype=object)

In [33]:
data_app['Reviews'] = data_app['Reviews'].astype(int)
data_app['Reviews'].dtype

dtype('int32')

Removing duplicated rows

In [36]:
duplicated = data_app.duplicated(subset=['App', 'Category', 'Type'], keep='first')
data_app[duplicated].sort_values(by=['App'])

App            Category  \
1407                              10 Best Foods for You  HEALTH_AND_FITNESS   
2543                         1800 Contacts - Lens Store             MEDICAL   
2385                         2017 EMRA Antibiotic Guide             MEDICAL   
1434                       21-Day Meditation Experience  HEALTH_AND_FITNESS   
5415                            365Scores - Live Scores              SPORTS   
...                                                 ...                 ...   
3085  theScore: Live Sports Scores, News, Stats & Vi...              SPORTS   
3014  theScore: Live Sports Scores, News, Stats & Vi...              SPORTS   
3202                           trivago: Hotels & Travel    TRAVEL_AND_LOCAL   
3118                           trivago: Hotels & Travel    TRAVEL_AND_LOCAL   
8291                     wetter.com - Weather and Radar             WEATHER   

      Rating  Reviews     Size  Installs  Type  Price Content Rating  \
1407     4.0     2490   3800.0    500000  Free   0.00   Everyone 10+   
2543     4.7    23160  26000.0   1000000  Free   0.00       Everyone   
2385     4.4       12   3800.0      1000  Paid  16.99       Everyone   
1434     4.4    11506  15000.0    100000  Free   0.00       Everyone   
5415     4.6   666246  25000.0  10000000  Free   0.00       Everyone   
...      ...      ...      ...       ...   ...    ...            ...   
3085     4.4   133833  34000.0  10000000  Free   0.00   Everyone 10+   
3014     4.4   133825  34000.0  10000000  Free   0.00   Everyone 10+   
3202     4.2   219848  14000.0  50000000  Free   0.00       Everyone   
3118     4.2   219848   7600.0  50000000  Free   0.00       Everyone   
8291     4.2   189310  38000.0  10000000  Free   0.00       Everyone   

                Genres       Last Updated         Current Ver  \
1407  Health & Fitness  February 17, 2017                 1.9   
2543           Medical      July 27, 2018               7.4.1   
2385           Medical   January 27, 2017               1.0.5   
1434  Health & Fitness     August 2, 2018               3.0.0   
5415            Sports      July 29, 2018               5.5.9   
...                ...                ...                 ...   
3085            Sports      July 25, 2018              6.17.2   
3014            Sports      July 25, 2018              6.17.2   
3202    Travel & Local     August 2, 2018  Varies with device   
3118    Travel & Local     August 2, 2018  Varies with device   
8291           Weather     August 6, 2018  Varies with device   

             Android Ver  
1407        2.3.3 and up  
2543          5.0 and up  
2385        4.0.3 and up  
1434          4.1 and up  
5415          4.1 and up  
...                  ...  
3085          4.4 and up  
3014          4.4 and up  
3202  Varies with device  
3118  Varies with device  
8291  Varies with device  

[1095 rows x 13 columns]

In [40]:
data_app.drop_duplicates(subset=['App', 'Category', 'Type'] , keep='first' , inplace=True)

In [41]:
#export data to Edepcion_googleplaystore.csv
data_app.to_csv("C:/Users/mneme/Desktop/Edepcion_Application.csv", index = False)